File: modules/expo-bluesky-oauth-client/ios/ExpoBlueskyOAuthClientModule.swift
import ExpoModulesCore
import JOSESwift

public class ExpoBlueskyOAuthClientModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBlueskyOAuthClient")

    AsyncFunction("digest") { (data: Data, algorithmName: String, promise: Promise) in
      if algorithmName != "sha256" {
        promise.reject("Error", "Algorithim not supported")
        return
      }
      promise.resolve(CryptoUtil.digest(data: data))
    }

    // We are going to leave this as sync to line up the APIs. It's fast, so not a big deal.
    Function("getRandomValues") { (byteLength: Int) in
      return CryptoUtil.getRandomValues(byteLength: byteLength)
    }

    AsyncFunction ("generateJwk") { (algo: String?, promise: Promise) in
      if algo != "ES256" {
        promise.reject("GenerateKeyError", "Algorithim not supported.")
        return
      }
      
      let keypair = try? CryptoUtil.generateKeyPair()

      guard keypair != nil else {
        promise.reject("GenerateKeyError", "Error generating JWK.")
        return
      }

      promise.resolve(keypair)
    }

    AsyncFunction("createJwt") { (header: String, payload: String, jwk: String, promise: Promise) in
      guard let jwt = JWTUtil.createJwt(header: header, payload: payload, jwk: jwk) else {
        promise.reject("JWTError", "Error creating JWT.")
        return
      }
      promise.resolve(jwt)
    }

    AsyncFunction("verifyJwt") { (token: String, jwk: String, promise: Promise) in
      promise.resolve(JWTUtil.verifyJwt(token: token, jwk: jwk))
    }
  }
}

File: modules/expo-bluesky-oauth-client/android/src/main/java/expo/modules/blueskyoauthclient/ExpoBlueskyOAuthClientModule.kt
package expo.modules.blueskyoauthclient

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoBlueskyOAuthClientModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBlueskyOAuthClient")

    AsyncFunction("digest") { value: ByteArray, algorithmName: String ->
      if(algorithmName != "sha256") throw Exception("Unsupported algorithm")
      return@AsyncFunction CryptoUtil().digest(value)
    }

    Function("getRandomValues") { byteLength: Int ->
      return@Function CryptoUtil().getRandomValues(byteLength)
    }

    AsyncFunction("generateJwk") { algorithim: String? ->
      if (algorithim != "ES256") {
        throw Exception("Unsupported algorithm")
      }
      return@AsyncFunction CryptoUtil().generateKeyPair()
    }

    AsyncFunction("createJwt") { header: String, payload: String, jwk: String ->
      return@AsyncFunction JWTUtil().createJwt(header, payload, jwk)
    }

    AsyncFunction("verifyJwt") { token: String, jwk: JWK ->
      return@AsyncFunction JWTUtil().verifyJwt(token, jwk)
    }
  }
}
